---
layout: post
author: pzc
title: Go的defer实现
tags: [Go]
date: 2022-07-03 00:00 +0800
toc:  true
---

## defer作用

协程g中会维护了一个链表_defer，在函数返回前可能会通过调用deferreturn来遍历调用defer的操作。如果发生了panic，也会遍历g中的_defer链表，除非defer顶层函数中调用了recover成功恢复，否则最终会终止程序。    
根据情况，编译器选择会在栈（deferprocStack）或者堆中（deferproc）为defer分配内存（待分析）。新创建的defer会插入到g的链表_defer中。defer中的字段pc通过getcallerpc获取，字段sp通过getcallersp获取，如果发生panic后recover后，可以通过这俩恢复。

## panic后的recover实现
{% highlight js %}
func main() {
  test()
  fmt.Println("a")
}
func test(){
  defer func () {
    panic("a")
  }()
  fmt.Println("test")
}
{% endhighlight %}
编译后部分汇编代码如下：
![图1](https://raw.githubusercontent.com/1053226965/blog/main/images/Snipaste_2022-07-03_17-41-01.png)

![图2](https://raw.githubusercontent.com/1053226965/blog/main/images/Snipaste_2022-07-03_17-41-20.png)

{% highlight js %}
// 栈上创建defer时，会调用这个函数，参考上图的汇编代码
func deferprocStack(d *_defer) {
  ...
  d.sp = getcallersp() // 发起defer的函数的sp
  d.pc = getcallerpc() // 调用者调用deferprocStack的函数返回后的pc
  ...
  return0() // 返回0给调用者后，执行TEST指令后，ZF标志位为零，通过JMP调转到下一条指令继续执行
}

// 触发panic后调用
func gopanic(e any) {
  ...
  //如果在defer顶层函数中调用了recover，并且可以恢复成功
  if p.recovered {
    ...
    mcall(recovery) // 切换到系统栈，调用recovery
    ...
  }
  ...
}

func recovery(gp *g) {
  ...
  gp.sched.sp = sp // defer中的sp
  gp.sched.pc = pc // defer中的pc
  gp.sched.lr = 0
  gp.sched.ret = 1 // 返回1给调用者后，执行TEST指令后，ZF标志位为零，通过JNE跳转到结尾处调用deferreturn
  gogo(&gp.sched)
}
{% endhighlight %}

根据返回值不同，区分是新建defer时的返回还是recover后的返回。如果是新建defer时的返回（return0），跳转到吓一条指令；如果是recover后的返回（return1），跳转到deferreturn处。非常巧妙。